### DISCUSSION TITLE
potential context processing levenstien culling optimization

### DISCUSSION THREAD

**Discussion by Swight1423** [ggml-org/llama.cpp#3956] [Ideas]
I have been messing with a personal deterministic model architecture and came up with a potential context optimization though I don't know the availability of the required info in traditional models or if this is already done. This is based on in my head theory crafting I haven't tried implementing it yet.
Requirements:
Knowledge of the tokens following the previous token in the dataset
the levenstien distance for the entire context for the top scoring(lowest to highest) tokens
variable length vocabulary
Process:
find the difference in the levenstien distance between the top scoring tokens
if the next token doesn't beat this difference it means that the next token must be one of the tokens following the previous token in the dataset.
with this knowledge we can reduce the number of tokens to process to those that are long enough to produce a levenstien distance score of at least that difference. 
you could further reduce the number to process by trimming the beginning and end of each token where it matches the context since those characters would not influence the levenstien distance. 
(the rest is much more theory since I can't fully think through it)
if you do find tokens with a bigger levenstien distance after calculation keep track of the largest one. 
my guess would be that you would be looking at the tokens that follow the token with the smallest levenstien distance that is bigger in the dataset.

**Comment by ggerganov** [comment]
I'm finding it hard to follow the description. Would be better if you provide a specific example.

**Reply by Swight1423** [reply]
I have a hard time following it myself and I came up with it so I understand. here is a step by step example to the best of my ability. no guarantees I got things perfect. 
lets say you have a context with "12345" from a previous generation and you want to generate next Context[6]. 
1.get the levenstien distance for Context[1-4] for each token 
2.sort the levenstien distances from step 1.
3. find the difference between the lowest two of the sorted values.
4. the value in step 3 is the how much difference in levenstien distance is required to make the result not be one that follows the context[5] in the dataset.
5. to create a given levenstien distance you need to have at least that number of characters in one of the strings involved 
6. so we can remove any token from current consideration that isn't greater than or equal the value in step 3.
7. for the remaining tokens from step 6 compare the beginning and end to Context[5] and remove any characters that are the same as these do not effect levenstien distance.
8. remove any token from step 7 that isn't greater than or equal in length to the value in step 3
9. if none remain Context[6] follows an instance of Context[5] in the dataset
10. if some remain find the highest levenstien score of the set remember need to remove what you did in step 7 from both the tokens and Context[5] during the calculation. 
11. the removed portions in step 7 and 10 can be restored now. the removal was only to facilitate early removal of token options and improve the calculation speed in step 10 since levenstien distance performance is highly dependent on the length of the strings compared.
12. if all the scores are less than or equal to step 3 value Context[6] follows an instance of Context[5]
13. else
14. Context[6] follows the token from step 2 with the lowest score that is higher than the value from step 10.

**Reply by KerfuffleV2** [reply]
> so we can remove any token from current consideration that isn't greater than or equal the value in step 3.

I think you might be assuming if we could remove tokens from consideration like that, then we could calculate the next result more efficiently? Unfortunately, it doesn't really work like that. Let's say the model has a vocab size of 32,000 and we can say for sure the next token it generates absolutely will not be in the range 15,999 through 31,999. As far as I know, there's nothing you could really do with that information to make the calculation faster.

Also, I'm not sure if you already know but once you have run the model, the result you get back from it is a giant array of numbers equal to the vocab size, where each item is how probable the model thinks that token is. So for the hypothetical 32,000 vocab model, you get back 32,000 probabilities. It doesn't matter which were eliminated as possible before you ran the model, you'll still get back probabilities equal to the vocab size.

I still don't see what Levenshtein distance has to do with this or how it can help, but even if  we just took that as a given I don't see what you could do with that information.

However, the source is in this repo and there are also high level bindings if you don't know C/C++. You can try to implement what you're talking about and if you're successful then the results will speak for themselves.

**Reply by Swight1423** [reply]
if my understanding is right not sure that is quite accurate the number of vocab returned for a given character should have an upper bound based on how often it occurs in the dataset could be lower if some of the duplicates are followed by the same token. so for some characters this may not exceed the vocab limit. though it is possible or even likely that others would and demonstrate your scenario. My personal model(which I am developing in vb.net may implement this for that once I sort my software for generating the models) which I came up with this for is kinda a unique case because it has far fewer probabilities to deal with. I thought would bring it up here just in case it was useful outside of my use case.

**Reply by Swight1423** [reply]
most of my knowledge of levenstien distance is from https://www.jetbrains.com/dotnet/guide/tutorials/dotnet-days-online-2022/maximising-algorithm-performance-in-dotnet-levenshtein-distance/ which talks about ways of optimizing the algorithm.

**Reply by KerfuffleV2** [reply]
> if my understanding is right not sure that is quite accurate the number of vocab returned for a given character should have an upper bound based on how often it occurs in the dataset c

By "character", you mean token? Anyway, unfortunately that's not correct. You always get back probabilities equal to the size of the vocab.

Since you said you're using VB, perhaps you're using a higher level interface that's obscuring those types of details. For example, if you look at the probabilities after sampling such as Top-K, Top-P, etc then the list will be much, much shorter.

But the model still produced `n_vocab` probs after evaluation.

**Reply by Swight1423** [reply]
I have actually been trying to write my own tokenizers for the unique needs of my models since I am trying to run experiments with token de-duplication.

**Reply by Swight1423** [reply]
so far I haven't really touched the samplers at all.

**Comment by KerfuffleV2** [comment]
> I have been messing with a personal deterministic model architecture 

Aren't they all deterministic?

> if the next token doesn't beat this difference it means that the next token must be one of the tokens following 

If you know whether the next token beats the difference, then don't you already know the next token?

Unfortunately, the description is pretty hard to follow. Are you trying to say something like given certain tokens it's basically guaranteed that other ones will follow. For example, like if you have ` guarant` then `eed` is basically the only other possibility so you can just feed it to the model instead of having the model actually generate it? The idea being to only make the model generate when there are real alternatives to "think" about.

**Reply by Swight1423** [reply]
I am saying it is possible to determine the tokens that have a probability of changing the outcome without calculating all of them.

**Reply by Swight1423** [reply]
and the highest of the calculated levenstien distances of that set if over the threshold may have a possibility of determining the actual next token the correct value follows but I am less sure of that part.

**Reply by Swight1423** [reply]
what I meant by deterministic: in my personal experimental model I am developing every token is unique and thus always followed by only one specific token in the dataset. my vocabulary is actually an index into the original dataset. my understanding is a lot of models each token has a percentage chance of being followed by one of a selection of tokens because you have a bunch of duplicates the tokens represent which may be followed by different values.
